File: test-server/app/src/main/java/com/polis/app/api/CourseService.kt
package com.polis.app.api

import com.polis.app.model.*
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.POST

interface CourseService {
    
    @POST("upsertCourse")
    suspend fun upsertCourse(@Body course: CourseDto): Response<RespSingleDto<CourseDto>>
    
    @POST("filterCourses")
    suspend fun filterCourses(@Body filter: SimpleStringFilterDto): Response<RespSliceDto<CourseDto>>
    
    @POST("deleteCourse")
    suspend fun deleteCourse(@Body request: LongIdDto): Response<RespSingleDto<CourseDto>>
    
    @POST("getCourse")
    suspend fun getCourse(@Body request: LongIdDto): Response<RespSingleDto<CourseDto>>
    
    @POST("associateTeacherToCourse")
    suspend fun associateTeacherToCourse(@Body request: CourseTeacherAssocDto): Response<RespSingleDto<CourseDto>>
    
    @POST("removeTeacherFromCourse")
    suspend fun removeTeacherFromCourse(@Body request: CourseTeacherAssocDto): Response<RespSingleDto<CourseDto>>
} 

File: polis-frontend/src/app/services/course.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { 
  CourseDto, 
  SimpleStringFilterDto, 
  LongIdDto, 
  RespSingleDto, 
  RespSliceDto,
  CourseStudentAssocDto
} from '../models/dto.types';

@Injectable({
  providedIn: 'root'
})
export class CourseService extends ApiService {

  // Get all courses with pagination and filtering
  filterCourses(filter: SimpleStringFilterDto): Observable<RespSliceDto<CourseDto>> {
    return this.post<RespSliceDto<CourseDto>>('/course/filter', filter);
  }

  // Get a single course by ID
  getCourse(request: LongIdDto): Observable<RespSingleDto<CourseDto>> {
    return this.post<RespSingleDto<CourseDto>>('/course/get', request);
  }

  // Create or update a course
  upsertCourse(course: CourseDto): Observable<RespSingleDto<CourseDto>> {
    return this.post<RespSingleDto<CourseDto>>('/course/upsert', course);
  }

  // Delete a course
  deleteCourse(id: number): Observable<RespSingleDto<void>> {
    return this.delete<RespSingleDto<void>>(`/course/${id}`);
  }

  // Associate student to course
  associateStudentToCourse(assoc: CourseStudentAssocDto): Observable<RespSingleDto<void>> {
    return this.post<RespSingleDto<void>>('/associateStudentToCourse', assoc);
  }

  // Remove student from course
  removeStudentFromCourse(assoc: CourseStudentAssocDto): Observable<RespSingleDto<void>> {
    return this.post<RespSingleDto<void>>('/removeStudentFromCourse', assoc);
  }
}

